src/main/java/com/minecolonies/coremod/colony/ColonyManager.java

```diff
@@ -992,58 +1008,155 @@ public static void onWorldLoad(@NotNull final World world)
+            {
+                @NotNull final File saveDir =
+                  new File(DimensionManager.getWorld(0).getSaveHandler().getWorldDirectory(), FILENAME_MINECOLONIES_PATH); //todo remove hardcoded world 0
+                final ZipOutputStream zos = new ZipOutputStream(fos);
+
+                for (final Colony backupTargetColony : colonies)
                 {
-                    addToZipFile(String.format(FILENAME_COLONY, i), zos, saveDir);
+                    @NotNull final File file = new File(saveDir, String.format(FILENAME_COLONY, backupTargetColony.getID()));
+                    if (file.exists())
+                    {
+                        addToZipFile(file, zos);
+                    }
                 }
-            }
-            addToZipFile(getSaveLocation().getName(), zos, saveDir);
+                addToZipFile(getSaveLocation(), zos);
 
-            zos.close();
-            fos.close();
+                zos.close();
+            }
+            catch (final Exception e)
+            {
+                /**
+                 * Intentionally not being propagated.
+                 */
+                Log.getLogger().warn("Unable to backup colony data, please contact an administrator");
+                Log.getLogger().debug("Exception:", e);
+                // Don't leave a mess - release the token before we return
+                backupInProgress = false;
+                return false;
+            }
+            // remove any surplus backups
+            pruneOldBackups(Configurations.administration.colonyBackupsToKeep);
+            // Schedule the next backup
+            scheduleColonyBackup();
+            // Release the token before we return
+            // no need to be thread safe here, we're the only ones who have it
+            backupInProgress = false;
+
+            return true;
         }
-        catch (final Exception e)
+        else
         {
-            /**
-             * Intentionally not being thrown.
-             */
-            Log.getLogger().warn("Unable to backup colony data, please contact an administrator");
+            // We were denied the backup token. That means there's
+            // another thread already in the process of running the backup
             return false;
         }
-
-        return true;
     }
 
-    public static void addToZipFile(final String fileName, final ZipOutputStream zos, final File folder)
+    public static void addToZipFile(final File file, final ZipOutputStream zos)
     {
-        final File file = new File(folder, fileName);
-        try(FileInputStream fis = new FileInputStream(file))
+        try (FileInputStream fis = new FileInputStream(file))
         {
-            zos.putNextEntry(new ZipEntry(fileName));
+            zos.putNextEntry(new ZipEntry(file.getName()));
             Files.copy(file, zos);
-            fis.close();
         }
         catch (final Exception e)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)